4.1.cpp

```diff
@@ -0,0 +1,201 @@
+
+
+
+
+int main ()
+{
+  const int min_value = -15;
+  const int max_value = 15;
+  unsigned int n;
+  std::cout << "enter the dimension of the array-";
+  std::cin >> n;
+  int *my_array = new int[n];
+  std::cout << "do you want to enter numbers manually(1) or fill in randomly(0)?"<< std::endl;
+  int b;
+  while (true)
+    {
+      std::cin >> b;
+      switch (b)
+	{
+	case 1:
+	  input_manual (my_array, n, min_value, max_value);
+	  break;
+	case 0:
+	  input_random (my_array, n, min_value, max_value);
+	  break;
+	default:
+	  std::cout << "you can only enter 1 or 0 " << std::endl;
+	  continue;
+	}
+      break;
+    }
+  std::cout << "got an array :" << std::endl;
+  output_array (my_array, n);
+  int result = function1 (my_array, n);	//the product of all even numbers from -15 to 15
+  if (result == 1)
+    {
+      std::cout << "there are no even numbers in your array  " << std::endl;
+    }
+  else
+    {
+      std::cout << "the product of all even numbers from -15 to 15 is equal to " << result << std::endl;
+    }
+  function2 (my_array, n);	//replace all elements of the array with even numbers by the square of their number
+  std::cout << " array after replacement : " << std::endl;
+  output_array (my_array, n);	//array output
+  int k;
+  result = function3 (my_array, n, &k);	//numbers divisible by k with a remainder of 2
+  if (result == 0)
+    {
+      std::cout <<"there are no numbers in the array that give the remainder of 2 branches on "<< k << std::endl;
+    }
+  else
+    {
+      std::cout << "in the array " << result << " numbers divisible by " << k <<" with a remainder of 2" << std::endl;
+    }
+  delete[]my_array;
+}
+
+void input_manual (int *my_array, unsigned int n, int min_value, int max_value)
+{
+  std::cout << "enter " << n << " numbers from  " << min_value << " before " <<max_value << std::endl;
+  for (unsigned int index = 0; index < n; ++index)
+    {
+      while (true)
+	{
+	  std::cin >> my_array[index];
+	  if (my_array[index] >= min_value && my_array[index] <= max_value)
+	    {
+	      break;
+	    }
+	  else
+	    {
+	      std::cout << " repeat the input " << std::endl;
+	    }
+	}
+    }
+}
+
+void input_random (int *my_array, unsigned int n, int min_value, int max_value)
+{
+  for (unsigned int index = 0; index < n; ++index)
+    {
+      my_array[index] = rand () % 31 - 15;
+    }
+}
+
+ void output_array (int *my_array, unsigned int n)
+{
+  for (unsigned int index = 0; index < n; ++index)
+    {
+      std::cout << my_array[index] << std::endl;
+    }
+}
+
+ int function1 (int *my_array, unsigned int n)	//the product of all even numbers from -15 to 15
+{
+  int c = 1;
+  for (unsigned int index = 0; index < n; ++index)
+    {
+      if (my_array[index] % 2 == 0)
+	{
+	  c *= my_array[index];
+	}
+
+    }
+  return c;
+}
+
+void function2 (int *my_array, unsigned int n)	//replace all elements of the array with even numbers by the square of their number
```

Comment:
Если параметры функции не изменяются, то они явно должны быть обозначены `const`
[By Alex Allain](https://www.cprogramming.com/tutorial/const_correctness.html)
>The const keyword allows you to specify whether or not a variable is modifiable. You can use const to prevent modifications to variables and const pointers and const references prevent changing the data pointed to (or referenced).
>Const дает вам возможность более четко документировать вашу программу и фактически обеспечивать соблюдение этой документации. Ключевое слово const обеспечивает вашим пользователям гарантии, позволяющие оптимизировать производительность без угрозы повреждения их данных, обеспечивая соблюдение вашей документации. Например, константные ссылки позволяют указать, что упомянутые данные не будут изменены; это означает, что вы можете использовать константные ссылки как простой и немедленный способ повышения производительности для любой функции, которая в настоящее время принимает объекты по значению, не беспокоясь о том, что ваша функция может изменить данные. Даже если это произойдет, компилятор предотвратит компиляцию кода и предупредит вас о проблеме. С другой стороны, если бы вы не использовали константные ссылки, у вас не было бы простого способа гарантировать, что ваши данные не были изменены.